src/auth/auth.service.ts

```diff
@@ -0,0 +1,44 @@
+import { Injectable } from '@nestjs/common'
+import { JwtService } from '@nestjs/jwt'
+
+import { BadRequestError, NotFoundError } from '../common/errors/errors'
+import { compare } from '../common/utils/crypto.utils'
+import { UserService } from '../user/user.service'
+import type { User } from '../user/schemas/user.schema'
+
+@Injectable()
+export class AuthService {
+  constructor(
+    private readonly jwtService: JwtService,
+    private readonly userService: UserService,
+  ) {}
+
+  async getValidatedUser(
+    username: string,
+    pass: string,
+  ): Promise<Omit<User, 'password'>> {
+    // TODO: add `username` and `pass` fields validation using class-validator
+
+    const user = await this.userService.findUserByUsername(username)
+
+    // TODO: discuss if errors throwing should be duplicated here (the same handled in `findUserByUsername`)
+    if (!user) {
+      throw new NotFoundError(`User with username ${username} not found`)
+    }
+
+    const isPasswordMatch = await compare(pass, user.password)
+
+    if (!isPasswordMatch) {
+      throw new BadRequestError('Wrong password')
```

Comment:
Please do not use HTTP-layer exceptions in Service layer. Errors here should reflect your business logic and know nothing about requests.

Reply:
It was a custom `BadRequestError`, not a HTTP exception, but now it's replaced with custom `ValidationError` here
```
if (!isPasswordMatch) {
      throw new ValidationError('Wrong password')
}
```